### ISSUE TITLE
pl-order-blocks exception with ordered grading and indentation

### ISSUE THREAD

**Issue by nealterrell** [PrairieLearn/PrairieLearn#4792]
I have this very simple ordered blocks question:

```html
<pl-order-blocks answers-name="answer" grading-method="ordered" indentation="true">
    <pl-answer correct="true" indent="0">Place this block first.</pl-answer>
    <pl-answer correct="true" indent="0">Place this block second.</pl-answer>
    <pl-answer correct="true" indent="1">Place this block third. Indent it once.</pl-answer>
    <pl-answer correct="false" indent="0">Unused block</pl-answer>
    <pl-answer correct="false" indent="0">Incorrect block</pl-answer>
</pl-order-blocks>
```

It grades correctly if the student only uses up to three blocks in their answer. However, if they use more than 3 blocks, we get the following error:

```
Traceback (most recent call last):
  File "/PL/CODE-pldev/lib/python-caller-trampoline.py", line 175, in <module>
    worker_loop()
  File "/PL/CODE-pldev/lib/python-caller-trampoline.py", line 121, in worker_loop
    val = method(*args)
  File "/PL/DATA-pldev/courses/csulb-cecs174/elements/pl-order-blocks/pl-order-blocks.py", line 569, in grade
    if true_answer_indent[i] == "-1" or int(indent) == true_answer_indent[i]:
IndexError: list index out of range
Traceback (most recent call last):
  File "/PL/CODE-pldev/lib/python-caller-trampoline.py", line 187, in <module>
    raise Exception('worker process exited unexpectedly with status %d' % status)
Exception: worker process exited unexpectedly with status 256
```

(The path shows my local repository instead of the PL source, because I copied the element's code from PL to my course due to the Berkeley server being out of date.) 

I believe the issue is that `true_answer_indent` is built only from the `correct="true"` blocks, so it has 3 entries; however, the statement throwing the error is in a loop over the list `student_answer`, which in the error case has 4 entries.

**Comment by mwest1066** [comment]
@SethPoulsen

**Comment by SethPoulsen** [comment]
Yikes, another case missed by our inadequate test suite. This is another case where I don't use this feature personally so I didn't notice. @nealterrell if you want to submit a fix again I'm happy to review it. 

I'll also make sure I keep this case in mind when I finally get around to augmenting the test suite the way I want to.

**Comment by nealterrell** [comment]
@SethPoulsen I guess my most important question is: should we even try to score indentation if the student did NOT use only the correct blocks in the correct order? 

If the correct answer is A B C, the student currently gets 0 points for A C B, regardless of indentation. Should they also get 0 points for A B C D? or D A B C? or A B D C? If the answer is yes, that's a 0; then the fix is easy. If `final_score` is 0 after grading the order of the blocks, then simply don't bother checking for indentation. What do you think?

**Comment by SethPoulsen** [comment]
Figuring out how to do partial credit in a sensible way for `pl-order-blocks` problems is actually really difficult (which is one of the reasons that I made it so there was no partial credit option when I created the `dag` grading mode). 

@Frozemint what is your opinion on this? Were you the one that originally wrote this portion of the code base? Are there other people whose opinions we should ask?

If the behavior hasn't been well defined as of yet, we can discuss here what should be done. 

To be honest, I don't really have a lot of "skin in the game" so to speak because as of right now, I don't use the `ordered` grading mode, or the `indent` feature (I primarily do mathematical proofs with the `dag` grading mode, and we don't care about indentation for those).

**Comment by SethPoulsen** [comment]
this thread on partial credit algorithms is also relevant: https://github.com/PrairieLearn/PrairieLearn/issues/4652

**Comment by nealterrell** [comment]
I can confirm that changing this line: https://github.com/PrairieLearn/PrairieLearn/blob/master/elements/pl-order-blocks/pl-order-blocks.py#L412

to
 
```python
if check_indentation and final_score > 0:
```

fixes this particular issue, by not attempting to score indentation if the student put the blocks in the wrong order period. I have not tested this with other order blocks questions, literally only the one shown above. It feels like it should be fine; if `final_score` is 0, then [line 420](https://github.com/PrairieLearn/PrairieLearn/blob/master/elements/pl-order-blocks/pl-order-blocks.py#L420) is pointless anyways, and we will avoid the loop that is causing this issue. This will have to be expanded if #4652 ever gets traction, because we could then end up with the student's answer having a different number of blocks than the expected answer.

**Comment by Frozemint** [comment]
> Were you the one that originally wrote this portion of the code base?

Yep, I was the one who wrote most of the original code when we started pl-order-blocks to merge the old parsons element. 

> Are there other people whose opinions we should ask?

Honestly, I think we should ask and listen to the profs who are using the element. Most (if not all) of the current partial credit options are created out of necessity; profs started having really specific needs or requirements for giving partial credits in the original questions they came up with when using pl-order-blocks,  and that's how most of the partial credit options were implemented.
